pipeline/src/pipeline.rs

```diff
@@ -827,6 +828,12 @@ impl<T: FieldElement> Pipeline<T> {
         Ok(self.artifact.witness.as_ref().unwrap().clone())
     }
 
+    pub fn publics(&self) -> Result<Vec<(String, T)>, Vec<String>> {
+        let pil = self.optimized_pil()?;
+        let witness = self.witness()?;
```

Comment:
witness has optimized_pil as dependency, so if you assume the witness exists then optimized_pil should exist too